### DISCUSSION TITLE
Add visual offset option

### DISCUSSION THREAD

**Discussion by d-prob** [ppy/osu#13143] [Ideas] [answered]
Its just impossible to get good accuracy with a weird hit offset.
I would like to request a "hit offset" slider on the options menu.

**Comment by bdach** [comment]
"Audio offset" is already a thing in the settings; per-beatmap offset is also planned (see #8787). Does this answer your request?

**Comment by d-prob** [comment]
Audio offset is a bit helpful, but not that much since it just doesnt work sometimes because its just the keypress that is delayed, not the music.

**Comment by bdach** [comment]
That sounds wrong. Currently input is processed at 1000fps when multithreaded, so you shouldn't be seeing any delay. Are you using wired peripherals? Note that wireless devices can have intrinsic input delay due to having to transmit the inputs over the air.

I'm pretty sure an input offset is not something that will happen.

**Comment by abraker95** [comment]
Input offset sounds wrong. What is needed is visual offset.

**Comment by bdach** [comment]
I'm not sure I follow what this issue is asking for if that is the case. How would that even work?

**Comment by abraker95** [comment]
> How would that even work

Consider you are playing on a TV that has a 100 ms latency. Everything being displayed is displayed 100 ms late, so you hit everything 100 ms late. To fix that, the map visually needs to be offset to 100 ms earlier but the time you need to hit the notes needs to remain the same.

**Comment by bdach** [comment]
100ms late relative to what? If relative to audio, then you might as well put an audio offset of +100ms to compensate, no?

**Comment by abraker95** [comment]
100 ms late relative to actual timing of the note; the point in time the game expects the note to be hit.

**Comment by bdach** [comment]
...which depends on the audio timing, does it not? Unless you're expecting compensation for input lag? I'm really struggling to understand here.

**Comment by abraker95** [comment]
No, try play osu on a TV that has all the fancy HDR features enabled, for example. Audio latency wont suffer, but the TV will take time processing visual information, causing the display to be late relative to audio. 

So if you hit a key, you might hear audio 10 ms after you hit the key, but wont see it on the screen until 100 ms after you hit the key.

**Comment by Wieku** [comment]
@bdach That's a feature some VSRGs do have. They have two settings - Audio Offset and Visual Offset. In some cases audio may be in sync but if you're not solely dependent on music rhythm but visuals then display's reaction time (or your eyes OR when you have bad keyboard with 30ms+ processing delay) can be a problem.
It doesn't move the hitobject timing window like audio offset, just visual representation of it.

**Comment by Wieku** [comment]
To be easier to understand:
![image](https://user-images.githubusercontent.com/5307179/89736382-4dce2c00-da69-11ea-87a1-7306861c8c78.png)
We have a hitcircle here. It's supposed to be clicked at blue arrow. What visual offset does, it moves the visual representation of hitcircle to the red arrow, but it's timing stays at the blue arrow.

**Comment by bdach** [comment]
Okay, I *think* I get it - you're asking for the ability for the game to send frames earlier than audio to compensate for the round trip time of GPU -> screen. (Input would be unaffected by this.)

The "100ms" figure given sounds exaggerated though to the point to which that would probably be unplayable; I imagine that latency should be half of that at most which comes out to about 3-4 frames. Which sounds like not much to me but maybe at high levels it is significant. No idea.

**Comment by abraker95** [comment]
I tried playing mania on my Samsung 4k TV and was hitting 50s on all my hits. That felt like 100ms late. Yes, it was unplayable, but if I could change visual offset it can be playable.

**Comment by Wieku** [comment]
@bdach It's not exaggerated if we're talking about TVs or projectors. I think it will also positively affect mobiles which do have 80ms+ of touch delay on average.

**Comment by peppy** [comment]
First step is to set your tv to "game mode". As osu! relies on hitsounds being in time, a visual offset is not something which will feel good.

**Comment by abraker95** [comment]
I know some tvs have a game mode. I used the tv as a supposedly relatable example to make a point across. As @Wieku mentioned earlier, visual offset would be beneficial for mobile and is more commonly used to compensate for reaction time. I don't know what hitsounds have to do with visual offset; it should not do anything to hitsounds.

**Comment by peppy** [comment]
At the point a visual offset is required, the game is running on inadequate hardware. If you can't understand how it affects hitsounds, you may need to think about offsets for a few days. Maybe do some diagrams. Offsets aren't easy to understand so if you aren't willing to do the mental work, trust that what I say is correct.

**Comment by abraker95** [comment]
**Key:**
- Blue line: Timing point where note is displayed
- Red line: Timing point where hitsound is played/heard
- Purple line: Timing point where sound in music associated with the note plays
- Green circle: Timing point where the game logic interprets 0 ms hiterror

### Audio latency example
> **Player perspective:**
> Audio latency is in effect, causing the music to be heard some amount of time after when the note needs to be hit. The player hits the note when visible, but the hit does not match the music. The map appears off-sync with the music.
>
> **Game perspective:**
> Everything is in sync, but map needs to be moved to compensate for audio
>
> ![](https://i.imgur.com/6xnJ7wo.png)

### Audio latency example with correct audio offset
> **Player perspective:**
> Points when note needs to be hit and sound in music associated with the note are now in sync
>
> **Game perspective:**
> Audio latency is compensated by shifting all of the notes in the map to match the audio as heard by the player
>
> ![](https://i.imgur.com/aAxY8lW.png)

### Visual latency example with correct audio offset
> **Player perspective:**
> Visual latency is in effect, causing the note to appear later than it must actually be hit. While the timings of music and when the note needs to be hit do match, when the note appears and the player hits it, the hit is late.
>
> **Game perspective:**
> Everything is in sync, but notes need to be displayed a bit before the point when game interprets center of hit
>
> ![](https://i.imgur.com/DMkhXPN.png)

### Visual latency example with correct audio offset and correct visual offset
> **Player perspective:**
> Points when note needs to be hit and when the note appears are now in sync
>
> **Game perspective:**
> Visual latency is compensated by shifting the visual representation of the map so notes appear earlier than they really are
>
> ![](https://i.imgur.com/dfLBx56.png)

**Comment by peppy** [comment, marked as answer]
Let me attempt to set you in the right direction, but this will be my last reply (as I have already said, offsets are hard and I don't have the time to attempt to explain this for the nth time here – it's something you'll need to come to terms with):

You are proposing visuals are forced to appear sooner, which means the song is already at a point in the future. So what happens when the user hits? The hitsounds cannot be played in the past – the point where it needs to be played has already elapsed.

I'm sure there's a term for this, but let's go with "chronologically impossible". Other games that provide visual offsets only do so because they either

- don't have hitsounds
- silence hitsounds when in effect

The true solution here is to fix what is causing latency, not adjust to it.

**Comment by LastExceed** [comment]
Peppy has a point, but unfortunately he made a very confusing mistake: The problem in this case is not hit**sounds** but hit**animations**.

hit**sounds** become a problem when you change the **audio** offset. you can observe this in stable, when you let autoplay play a map with a global offset you will hear that the hitsounds don't line up with the music. This is why many players turn off either global offset or hitsounds entirely.

When using a visual offset you have the same problem but with the animations. To get an idea what that would look like, here's a video from StepMania: https://youtu.be/iaqs6uQg1JI

While the option does allow re-aligning the point in time where the display shows the notes reaching the receptors with the point in time where you hear the beat and need to press a button (assuming you have already synchronized those two with each other using audio offset), the receptor highlighting will still show up late (since we can't send the signal back in time), the notes will disappear too late, and long notes have this wonky teleport effect.

If you want to go to full circle we'd also need the third and last component "input offset" (shifting judgement windows in time), as the keyboard too can have input latency. However it is technically not necessary to have all 3 options (audio/video/input offset) as each one is effectively a combination of the other two.

I agree that one should always try to minimize latency rather than compensating it, but sometimes we're simply limited by the hardware we have. I have throughout the years experienced all three at some point:
* keyboards with awful input latency (eg my laptop's internal keyboard easily has 30ms+ input latency because cheap garbage)
* displays with long reaction times (unfortunately not every TV has a "gaming mode")
* speakers with high audio latency (eg my headset's audio latency increases by 12ms when i connect it over usb instead of aux)

Granted this usually means the hardware (or driver) sucks, but we can't reasonably expect the player to buy new hardware just for osu!. Compensating with side effects is still better than not being able to address the issue at all.

**Comment by d-prob** [comment]
I'm back and I think i found the solution to the first problem
https://youtu.be/PXoJIrueUGY
and for the visual offset thing:
the notes falling will still be the same, but displays the score once after (or on) the keypress.
i dont know if i explained it correctly :/

**Comment by LastExceed** [comment]
The video shows a strong visual delay, but it seems to be constant (increasing the scroll speed doesnt change it, it just makes it more apparent). However visual delay is usually caused by the monitor and thus shouldn't be visible in a recording (unless you point a real camera to a screen , which you didn't). Idk what causes this, but this is definitely something that needs to be fixed rather than compensated.

Playing in windowed mode can cause a bit of visual delay due to buffering in the dwm, but i shouldn't be this bad. Still, if you were playing windowed, try exclusive fullscreen just to be sure (and use auto-play for more accurate visuals)

**Comment by LukeConfait** [comment]
Not intending on reviving dead threads for no reason. Mania players are able to set a visual offset in the skin ini by changing the height of the judgement line but this becomes a pain to adjust especially in lazer at the moment.

It needs to be set per skin and adjusted when scroll speed is changed or else accuracy becomes heavily impacted.
In response to "The true solution here is to fix what is causing latency, not adjust to it.", I'm not needing it to fix any latency its just how I read the mania notes.

I don't know about the issue with hitsounds but I haven't seen an issue with them from manually adjusting the skin and personally I don't use hitsounds. 

If it doesn't get added as a feature we will be stuck with manually editing it in the skin ini. So at least it should be made easier to edit this.

**Reply by frenzibyte** [reply]
I would suggest opening a separate thread for this as this seems like something osu!(stable) users would expect to work and may suffer from, and is not relevant to adding a "Visual Offset" option as per the main idea of this discussion thread.

**Reply by LukeConfait** [reply]
Ill think about starting another thread then, My reason for replying here is that the option requested would essentially do the same thing but might be implemented differently or is the same option in other games.

It also seemed based on some of the replies that this was unlikely to be included at least for the way it was requested initially

**Reply by JSG-GIT** [reply]
It should be noted that for players who hit 'late', a visual offset by editing the ``hitposition`` in the ``skin.ini`` only does so much, as you cannot increase the value beyond 480. This workaround may not work for anyone who needs a value higher than that.

**Comment by sona-mini-studio** [comment]
This is trivial for most players since they have <20 ms total visual (physical input to motion picture) latency. It's easy to compensate such position offset (s=vt) by eyes.

In other environments, however, Visual Offset is a good option. I need the very _visual clue_ that I can get zero error value when I tap at the timestamp the picture shows. If there isn't Visual Offset, this would be rarely achieved where total visual latency is long. Typically Visual Offset is available on mobile platforms, whose total visual latency is long. In desktop computer world this also benefits players with low-end environment. They can use VSync to trade latency for stable frame timing in Taiko and Mania mode.

Visual Offset is nothing about reducing latency. (why I even need offsetting if all events are at the same timestamp?) Latency optimization is complexed engineering on other branches. Visual Offset just adjusts the position _where_ hitobjects show, making the _visual clue_ true. (while Audio Offset adjusts _when_ hitobjects show, affecting timing for judgement)

**Comment by Glebsin** [comment]
I have bluetooth keyboard and I want to play (osu!mania) with this keyboard but I can't do it too well without visual offset. It has recently become a revelation for me that I can play with this keyboard with near-zero timing distribution. 

After playing "Fortnite Festival" I saw the adjustment of both audio and visual offset. Then, after searching for information on the differences between these offsets, I learned that visual offset can help if you have high keyboard latency. I set the visual offset to -15ms and it helped. Then I decided to try vsrg "Quaver", there was also a visual offset setting, I set the same visual offset -15ms and it also helped. 

So it would be nice for me to see a visual offset setting in osu. The few ways I've seen to replace the visual offset settings are too complicated or unclear.

**Reply by peppy** [reply]
Try setting "universal offset" in settings. It should have the same effect.

**Reply by ItsCollector** [reply]
> Try setting "universal offset" in settings. It should have the same effect.

There's only audio offset in both stable and lazer, this isn't what commenter is asking for. These other VSRGs have a visual offset to account for hardware / game latency. It's very common for mania players to turn off the hit sounds anyway because they already sound off-sync with the music, which is why mania players may not agree with the answer to this thread. 

Changing the hit position is a band-aid fix as for players it might completely throw off their reading when using skins without receptors (like the popular "r-skin"). I personally feel some kind of weird latency when going from lazer back to stable (I have to practice there for MWC) and changing the hit position with my bar skin moves the judgement line asset down. I can't play with that change, it feels very bad to use.

**Reply by abraker95** [reply]
I think it's hard for peppy to understand this because he is coming from the angle of the non osu!mania gamemodes. Visual offset only makes sense in osu!mania because it doesn't have the concept of approach rates. If this was added globally, playing maps with AR10 can be played with visual offset that shows notes earlier, and that defeats the purpose of approach rate as a difficulty.

With that said, this would be best implemented as an osu!mania specific mod. Unfortunately other gamemodes will have to suffer from visual latency by design.

**Comment by cihe13375** [comment]
I read through this discussion and I think I agree with LastExceed's post (https://github.com/ppy/osu/discussions/13143#discussioncomment-791693). However, I want to point out that adding visual offset on the game level essentially allows manipulating AR, at least in std mode.

To see why, let's look at how it is intended to work (compensating for the latency of monitor). Assuming a 500ms of monitor latency/visual offset and a 500ms of approach time, it would look like this:
<img width="1959" height="1197" alt="1111" src="https://github.com/user-attachments/assets/ae8e576e-5d17-40a7-a9eb-934c7c6b6d13" />
The player hits the note at 0ms, but the note only disappear at 500ms. This is the consequence of visual offset, which is already explained in LastExceed's post.

OK maybe it just looks bad, we can accept it. But what if I set this 500ms visual offset while my monitor does not have any latency?
<img width="1971" height="1193" alt="2222" src="https://github.com/user-attachments/assets/6a7260ac-7b2e-4869-b94c-c19109760ba5" />
Now there is a 1000ms interval between the note's appearance and its hit! This gives the player an extra 500ms of reaction time and can be helpful when reading high AR.

OK, maybe you say that we should cut the initial approach circle animation to compensate for that. Then it becomes:
<img width="1995" height="1202" alt="3333" src="https://github.com/user-attachments/assets/f61cc3cc-81a3-4eaf-a8d6-95d831243102" />
Maybe it's fine. But what if I do have a high latency monitor?
<img width="1996" height="1209" alt="4444" src="https://github.com/user-attachments/assets/c9fa2819-90b7-4b56-a22a-647aea641887" />
Now the time between note appearing and player's hit is 0ms. Well, 0ms is probably not exploitable. However, the point is that the appear-to-hit interval can be decreased. This does not decrease the density of objects on the screen (since the notes disappear late now), but I still think it can be helpful in reading low AR (especially with HD).

Mania is unaffected by this problem, since "AR" can be configured by the player. I don't play mania, but apparently mania already have some skin settings to effectively set visual offsets, though they are not perfect. So I think the correct path is to improve mania skinning.

**Reply by sona-mini-studio** [reply]
In std mode, players use approach circles to (visually) read **accurate** timing. So we could apply Visual Offset **only** to these timing-sensitive visual elements. 

For example, in an AR9 (600ms) map with HD mod and 30ms (approach earlier) Visual Offset, the fading effect shouldn't be affected and the (first) approach circle should start at -600ms and end at -30ms (to the ideal hit).

**Reply by sona-mini-studio** [reply]
The allowed range should be limited. I've never seen any reasonable value beyond -50ms (approach later) to +100ms (approach earlier) in rhythm games running on modern devices.

**Comment by JSG-GIT** [comment]
I'm going to drop an example of what a pretty extreme visual offset looks like in a different 4Key VSRG (FlashFlashRevolution), in the hopes that this shows the _desired result_ of the proposed change. This play was set using roughly -133ms in both what osu! would consider Universal Offset, but also a **-133ms Visual Offset**
https://www.youtube.com/watch?v=GXw0YQWJzIY

Secondly, it should be noted that major 4K VSRG's tend to support visual offset (sometimes called judge offset), I'll list a few:
Etterna, FlashFlashRevolution, StepMania, Quaver, heck even modern DanceDanceRevolution machines support it, the thing the entire genre is based on.

_Lastly my experience as a player:_
In osu!mania it is impossible for me to get scores of equivalent difficulty precisely because visual offset is not supported. Once you have 10 years of eye-to-muscle memory on an offset like this it's impossible to get rid off. It has led to me not playing with friends that play osu!mania because I cannot compete with them despite outranking them skill-wise in any other 4K VSRG. Realistically I should be able to S or even SS 4 star maps, but I can't even find myself getting an A rank on 3 star maps. Because all my inputs are getting scored 200 due to my visual offset.
